pkg/logs/client/tcp/destination.go

```diff
@@ -29,97 +26,118 @@ type Destination struct {
-	if d.conn == nil {
-		var err error
-
-		// We work only if we have a started destination context
-		ctx := d.destinationsContext.Context()
-		if d.conn, err = d.connManager.NewConnection(ctx); err != nil {
-			// the connection manager is not meant to fail,
-			// this can happen only when the context is cancelled.
-			return err
+// Start reads from the input, transforms a message into a frame and sends it to a remote server,
+func (d *Destination) Start(input chan *message.Payload, output chan *message.Payload, isRetrying chan bool) (stopChan <-chan struct{}) {
+	stop := make(chan struct{})
+	go func() {
+		for payload := range input {
+			d.sendAndRetry(payload, output, isRetrying)
 		}
-		d.connCreationTime = time.Now()
-	}
+		d.updateRetryState(nil, isRetrying)
+		stop <- struct{}{}
+	}()
+	return stop
+}
 
-	metrics.BytesSent.Add(int64(len(payload)))
-	metrics.TlmBytesSent.Add(float64(len(payload)))
-	metrics.EncodedBytesSent.Add(int64(len(payload)))
-	metrics.TlmEncodedBytesSent.Add(float64(len(payload)))
+func (d *Destination) sendAndRetry(payload *message.Payload, output chan *message.Payload, isRetrying chan bool) {
+	for {
+		if d.conn == nil {
+			var err error
+
+			// We work only if we have a started destination context
+			ctx := d.destinationsContext.Context()
+			if d.conn, err = d.connManager.NewConnection(ctx); err != nil {
+				// the connection manager is not meant to fail,
+				// this can happen only when the context is cancelled.
+				d.incrementErrors(true)
+				return
+			}
+			d.connCreationTime = time.Now()
+		}
 
-	content := d.prefixer.apply(payload)
-	frame, err := d.delimiter.delimit(content)
-	if err != nil {
-		// the delimiter can fail when the payload can not be framed correctly.
-		return err
-	}
+		content := d.prefixer.apply(payload.Encoded)
+		frame, err := d.delimiter.delimit(content)
+		if err != nil {
+			// the delimiter can fail when the payload can not be framed correctly.
+			d.incrementErrors(true)
+			return
+		}
 
-	_, err = d.conn.Write(frame)
-	if err != nil {
-		d.connManager.CloseConnection(d.conn)
-		d.conn = nil
-		return client.NewRetryableError(err)
-	}
+		_, err = d.conn.Write(frame)
+		if err != nil {
+			d.connManager.CloseConnection(d.conn)
+			d.conn = nil
+
+			if d.shouldRetry {
+				d.updateRetryState(err, isRetrying)
+				d.incrementErrors(false)
+				// retry (will try to open a new connection)
+				continue
+			}
+			d.incrementErrors(true)
+		}
 
-	if d.connManager.ShouldReset(d.connCreationTime) {
-		log.Debug("Resetting TCP connection")
-		d.connManager.CloseConnection(d.conn)
-		d.conn = nil
-	}
+		d.updateRetryState(nil, isRetrying)
 
-	return nil
-}
+		metrics.LogsSent.Add(1)
+		metrics.TlmLogsSent.Inc()
+		metrics.BytesSent.Add(int64(len(payload.Encoded)))
+		metrics.TlmBytesSent.Add(float64(len(payload.Encoded)))
+		metrics.EncodedBytesSent.Add(int64(len(payload.Encoded)))
+		metrics.TlmEncodedBytesSent.Add(float64(len(payload.Encoded)))
```

Comment:
I think the conversation in https://github.com/DataDog/datadog-agent/pull/10162#discussion_r766962890 still might be valid here

Reply:
In the case of TCP there is no compression so encoded vs non-encoded bytes are always the same